pkg/processor/worker/worker.go

```diff
@@ -156,7 +156,23 @@ func (w *Worker) Terminate() error {
 }
 
 func (w *Worker) Drain() error {
-	if err := w.runtime.Drain(); err != nil {
+	drainDoneControlMessageChan := make(chan *controlcommunication.ControlMessage)
+	err := w.Subscribe(controlcommunication.DrainDoneMessageKind, drainDoneControlMessageChan)
+	if err != nil {
+		w.logger.ErrorWith("Failed to subscribe to drainDone control messages",
+			"error", err)
+	}
+
+	defer func() {
+		err := w.Unsubscribe(controlcommunication.DrainDoneMessageKind, drainDoneControlMessageChan)
+		if err != nil {
+			w.logger.ErrorWith("Failed to unsubscribe from drainDone control messages",
+				"error", err)
+		}
+		close(drainDoneControlMessageChan)
+	}()
+
+	if err := w.runtime.Drain(drainDoneControlMessageChan); err != nil {
 		return err
 	}
 	w.logger.DebugWith("Successfully drained worker", "workerIndex", w.index)
```

Comment:
I don't think this should be in the worker.
The whole process of signaling and waiting for draining to be done is done on the runtime.
Plus, Subscribing and Unsubscribing is already done via the runtime (e.g see `w.Subscribe`), no need to add complexity in an "upper" abstraction (the worker).

Reply:
@TomerShor agreed